### PR TITLE
Implement CheckConfig in RunnerFactory to make autodiscover check configs

### PR DESCRIPTION
**Pull request by vjsamuel**
After `autodiscover` was implemented to retry failed configs, it began to retry spinning up configs every 10s. This is great for configs that are actually valid as it might be a pipeline related issue. Configs that say dont have `hosts` on it will always fail. In such scenarios, we should check the validity of the config and if it is invalid, it should not be retried.

This PR, implements the missing `CheckConfig` API by moving it from `Adapter` interface into the `RunnerFactory` interface. It is implemented only for metricbeat right now. It needs to be enhanced to Filebeat once the Filebeat refactor is complete as there is a known memory leak on trying to simply trying to create an input and not start/stop it. 

In future, each module/input can expose an API which validates a config instead of trying to call the constructor which is not optimal.

### CODE REVIEW

**Comment by ruflin** [comment]
jenkins, test this

**Comment by ruflin** [comment]
jenkins, test this

**Comment by ruflin** [comment]
jenkins, test this

**Comment by vjsamuel** [comment]
@exekias i have added tests and updated existing tests as well

**Comment by exekias** [comment]
jenkins, test this please

File: libbeat/autodiscover/autodiscover_test.go [hunk 1]
```diff
@@ -289,6 +298,67 @@ func TestAutodiscoverHash(t *testing.T) {
...
+	// Test start event
+	eventBus.Publish(bus.Event{
+		"start": true,
+		"meta": common.MapStr{
+			"foo": "bar",
+		},
+	})
+
+	// Timed sleep to make sure that bus event is processed
+	time.Sleep(time.Second)
```
Comment (exekias) [inline libbeat/autodiscover/autodiscover_test.go]: maybe you can use the `wait` helper here?

Comment (vjsamuel) [reply libbeat/autodiscover/autodiscover_test.go]: @exekias done

**Comment by exekias** [comment]
jenkins, test this please

**Comment by exekias** [comment]
I'm backporting this one as it fixes some memory leaks

**Pull request merged**